### PR TITLE
fuzz testing

### LINKED ISSUE #38: Implement fuzz testing
Implement the tests to prevent security flaws that may be caused by invalid/unexpected inputs/data.

### PR DESCRIPTION
**Pull request by curiecrypt**
This PR implements fuzz testing for MuSig2.

Closes #38

### CODE REVIEW

File: tests/testmusig2.c [hunk 1]
```diff
@@ -71,11 +72,18 @@ MUSIG2_ERROR musig2_helper_verify(unsigned char *serialized_pubkey_list, unsigne
...
+    for (i = 0; i <  size; i++)
+        serialized_pubkey[i] = rand();
+}
+
 #include "functiontest.c"
 #include "serdetest.c"
 }
 
 int main(int argc, char* argv[]){
+    srand ((unsigned int) time (NULL));
```
Comment (iquerejeta) [inline tests/testmusig2.c:86]: What is this for?

File: tests/serdetest.c [hunk 2]
```diff
@@ -48,4 +49,213 @@ TEST (musig2, commitments_serialize_deserialize) {
...
+
+    switch(res){
+        case MUSIG2_ERR_PARSE_PK:
+            err = secp256k1_ec_pubkey_parse(ctx, &pubkey, &serialized_pubkey_list[fuzz_index], MUSIG2_PUBKEY_BYTES_COMPRESSED);
+            ASSERT_EQ(err, 0);
+            break;
+        case MUSIG2_ERR_AGGR_PK:
+            err = secp256k1_ec_pubkey_parse(ctx, &pubkey, &serialized_pubkey_list[fuzz_index], MUSIG2_PUBKEY_BYTES_COMPRESSED);
+            ASSERT_EQ(err, 1);
+            break;
```
Comment (iquerejeta) [inline tests/serdetest.c]: This case can fit within  `default` case.

File: tests/serdetest.c [hunk 3]
```diff
@@ -48,4 +49,213 @@ TEST (musig2, commitments_serialize_deserialize) {
...
+        case MUSIG2_INVALID:
+            int cmp_res;
+            musig2_aggr_pubkey aggr_pubkey, fuzz_aggr_pubkey;
+            assert(musig2_prepare_verifier(&aggr_pubkey, serialized_pubkey_list, NR_SIGNERS));
+            assert(secp256k1_xonly_pubkey_from_pubkey(ctx, &fuzz_aggr_pubkey, NULL, &mcs_list[0].mc.aggr_pubkey));
+            cmp_res = secp256k1_xonly_pubkey_cmp(ctx, &fuzz_aggr_pubkey, &aggr_pubkey);
+            ASSERT_NE(cmp_res, 0);
+            break;
+        default:
+            ASSERT_NE(err, MUSIG2_OK);
```
Comment (iquerejeta) [inline tests/serdetest.c]: These two cases can be considered in a single case, and you would check the expected behaviour, mainly that: 
1. deserialisation of the fuzzed public key is correct
2. `musig2_prepare_verifier` is correct
3. `musig2_verify` fails

File: tests/serdetest.c [hunk 4]
```diff
@@ -48,4 +49,213 @@ TEST (musig2, commitments_serialize_deserialize) {
...
+
+            err = musig2_helper_sign(mcs_list, mps, NR_SIGNERS);
+            ASSERT_EQ(err, MUSIG2_OK);
+
+            err = musig2_aggregate_partial_sig(mps, signature, NR_SIGNERS);
+            ASSERT_EQ(err, MUSIG2_OK);
+
+            err = musig2_helper_verify(serialized_pubkey_list, signature, MSG_1, MSG_1_LEN, NR_SIGNERS);
+            ASSERT_EQ(err, MUSIG2_INVALID);
+    }
```
Comment (iquerejeta) [inline tests/serdetest.c]: case `MUSIG2_ERR_AGGR_R` and `MUSIG2_OK` are subsets of `default`. You can merge them all into a single case.

File: tests/serdetest.c [hunk 5]
```diff
@@ -48,4 +49,213 @@ TEST (musig2, commitments_serialize_deserialize) {
...
+            ASSERT_EQ(err, 1);
+
+            err = musig2_helper_sign(mcs_list, mps, NR_SIGNERS);
+            ASSERT_EQ(err, MUSIG2_OK);
+
+            err = musig2_aggregate_partial_sig(mps, signature, NR_SIGNERS);
+            ASSERT_EQ(err, MUSIG2_OK);
+
+            err = musig2_helper_verify(serialized_pubkey_list, signature, MSG_1, MSG_1_LEN, NR_SIGNERS);
+            ASSERT_EQ(err, MUSIG2_INVALID);
```
Comment (iquerejeta) [inline tests/serdetest.c]: Here we are really only concerned about the serialisation of the public key. Not about the signature. We already check that in the test below. We can remove these checks, otherwise we have duplicate code/tests.
```suggestion
        default:
            err = secp256k1_ec_pubkey_parse(ctx, &pubkey, &serialized_pubkey_list[fuzz_index], MUSIG2_PUBKEY_BYTES_COMPRESSED);
            ASSERT_EQ(err, 1);
```

File: tests/serdetest.c [hunk 6]
```diff
@@ -48,4 +49,213 @@ TEST (musig2, commitments_serialize_deserialize) {
...
```
Comment (iquerejeta) [inline tests/serdetest.c]: Also, same as above, in this test we only really care about serialisation. For the (in)validity of the signature, we already have another test, so no need to duplicate code/checks here:
```suggestion
        default:
            err = secp256k1_ec_pubkey_parse(ctx, &pubkey, &serialized_batch_list[fuzz_index], MUSIG2_PUBKEY_BYTES_COMPRESSED);
            ASSERT_EQ(err, 1);
```

**Review by iquerejeta — APPROVED** [review]
Minor comments. Once addressed, it can be merged.

File: tests/testmusig2.c (hunk 1 above)
Comment (curiecrypt) [reply tests/testmusig2.c:86]: It seeds the RNG (`rand()`) that we use to set a random index to fuzz one of the public keys in the list.

Comment (iquerejeta) [reply tests/testmusig2.c:86]: Ok, sounds good. Looks good to me then!
